File: behavioral-design-pattern/visitor-pattern/visitor.go
// Visitor Pattern
//
// Category: Behavioral Design Pattern
//
// Intent:
//   Represent an operation to be performed on elements of an object structure
//   without changing the classes on which it operates.
//
// When to use:
//   - You need to perform many unrelated operations on an object structure
//   - The object structure rarely changes, but you often define new operations
//   - You want to avoid polluting element types with operation-specific logic
//
// Key Participants:
//   - Visitor: declares visit methods for each concrete element type
//   - ConcreteVisitor: implements each visit method with specific behavior
//   - Element: declares an accept(visitor) method
//   - ConcreteElement: implements accept() by calling the visitor's corresponding method
//
// Double Dispatch:
//   The visitor pattern uses double dispatch to invoke the correct operation.
//   First dispatch: the element calls Accept(visitor) -- polymorphic on the element type.
//   Second dispatch: inside Accept(), the element calls visitor.VisitX(self) -- polymorphic
//   on the visitor type. This two-step indirection lets the runtime resolve both the element
//   type AND the visitor type, achieving behavior that depends on both.

package main

import (
	"fmt"
	"strings"
)

// ---------------------------------------------------------------------------
// Element hierarchy
// ---------------------------------------------------------------------------

// Element is the base interface -- every document element must accept a visitor.
type Element interface {
	Accept(visitor ExportVisitor) string
}

type Paragraph struct {
	Text string
}

func (p *Paragraph) Accept(visitor ExportVisitor) string {
	// Second dispatch: visitor now knows the concrete element type
	return visitor.VisitParagraph(p)
}

type Image struct {
	URL     string
	AltText string
}

func (img *Image) Accept(visitor ExportVisitor) string {
	return visitor.VisitImage(img)
}

type Table struct {
	Headers []string
	Rows    [][]string
}

func (tbl *Table) Accept(visitor ExportVisitor) string {
	return visitor.VisitTable(tbl)
}

// ---------------------------------------------------------------------------
// Visitor hierarchy
// ---------------------------------------------------------------------------

// ExportVisitor defines the interface for each export format.
type ExportVisitor interface {
	VisitParagraph(p *Paragraph) string
	VisitImage(img *Image) string
	VisitTable(tbl *Table) string
}

// --- HtmlExporter ---

type HtmlExporter struct{}

func (h *HtmlExporter) VisitParagraph(p *Paragraph) string {
	return fmt.Sprintf("<p>%s</p>", p.Text)
}

func (h *HtmlExporter) VisitImage(img *Image) string {
	return fmt.Sprintf(`<img src="%s" alt="%s" />`, img.URL, img.AltText)
}

func (h *HtmlExporter) VisitTable(tbl *Table) string {
	var headerCells string
	for _, hdr := range tbl.Headers {
		headerCells += fmt.Sprintf("<th>%s</th>", hdr)
	}
	var rowsHTML string
	for _, row := range tbl.Rows {
		var cells string
		for _, c := range row {
			cells += fmt.Sprintf("<td>%s</td>", c)
		}
		rowsHTML += fmt.Sprintf("<tr>%s</tr>", cells)
	}
	return fmt.Sprintf("<table><tr>%s</tr>%s</table>", headerCells, rowsHTML)
}

// --- MarkdownExporter ---

type MarkdownExporter struct{}

func (m *MarkdownExporter) VisitParagraph(p *Paragraph) string {
	return p.Text
}

func (m *MarkdownExporter) VisitImage(img *Image) string {
	return fmt.Sprintf("![%s](%s)", img.AltText, img.URL)
}

func (m *MarkdownExporter) VisitTable(tbl *Table) string {
	headerLine := "| " + strings.Join(tbl.Headers, " | ") + " |"
	seps := make([]string, len(tbl.Headers))
	for i := range seps {
		seps[i] = "---"
	}
	sepLine := "| " + strings.Join(seps, " | ") + " |"

	var rowLines []string
	for _, row := range tbl.Rows {
		rowLines = append(rowLines, "| "+strings.Join(row, " | ")+" |")
	}
	return headerLine + "\n" + sepLine + "\n" + strings.Join(rowLines, "\n")
}

// --- PlainTextExporter ---

type PlainTextExporter struct{}

func (pt *PlainTextExporter) VisitParagraph(p *Paragraph) string {
	return p.Text
}

func (pt *PlainTextExporter) VisitImage(img *Image) string {
	return fmt.Sprintf("[Image: %s (%s)]", img.AltText, img.URL)
}

func (pt *PlainTextExporter) VisitTable(tbl *Table) string {
	// Calculate column widths for alignment
	colWidths := make([]int, len(tbl.Headers))
	for i, h := range tbl.Headers {
		colWidths[i] = len(h)
	}
	for _, row := range tbl.Rows {
		for i, cell := range row {
			if len(cell) > colWidths[i] {
				colWidths[i] = len(cell)
			}
		}
	}

	fmtRow := func(cells []string) string {
		parts := make([]string, len(cells))
		for i, c := range cells {
			parts[i] = fmt.Sprintf("%-*s", colWidths[i], c)
		}
		return strings.Join(parts, "  ")
	}

	var lines []string
	lines = append(lines, fmtRow(tbl.Headers))
	dashes := make([]string, len(colWidths))
	for i, w := range colWidths {
		dashes[i] = strings.Repeat("-", w)
	}
	lines = append(lines, strings.Join(dashes, "  "))
	for _, row := range tbl.Rows {
		lines = append(lines, fmtRow(row))
	}
	return strings.Join(lines, "\n")
}

// ---------------------------------------------------------------------------
// Helper: export entire document
// ---------------------------------------------------------------------------

func exportDocument(elements []Element, visitor ExportVisitor) string {
	parts := make([]string, len(elements))
	for i, el := range elements {
		parts[i] = el.Accept(visitor)
	}
	return strings.Join(parts, "\n")
}

// ---------------------------------------------------------------------------
// Main -- demonstrate exporting a document in three formats
// ---------------------------------------------------------------------------

func main() {
	document := []Element{
		&Paragraph{Text: "Welcome to the visitor pattern demo."},
		&Image{URL: "https://example.com/photo.png", AltText: "A sample photo"},
		&Table{
			Headers: []string{"Name", "Role"},
			Rows: [][]string{
				{"Alice", "Engineer"},
				{"Bob", "Designer"},
			},
		},
	}

	visitors := []struct {
		Label   string
		Visitor ExportVisitor
	}{
		{"HTML", &HtmlExporter{}},
		{"Markdown", &MarkdownExporter{}},
		{"Plain Text", &PlainTextExporter{}},
	}

	for _, v := range visitors {
		fmt.Printf("=== %s Export ===\n", v.Label)
		fmt.Println(exportDocument(document, v.Visitor))
		fmt.Println()
	}
}

File: behavioral-design-pattern/visitor-pattern/visitor.py
"""
Visitor Pattern

Category: Behavioral Design Pattern

Intent:
    Represent an operation to be performed on elements of an object structure
    without changing the classes on which it operates.

When to use:
    - You need to perform many unrelated operations on an object structure
    - The object structure rarely changes, but you often define new operations
    - You want to avoid polluting element classes with operation-specific logic

Key Participants:
    - Visitor: declares visit methods for each concrete element type
    - ConcreteVisitor: implements each visit method with specific behavior
    - Element: declares an accept(visitor) method
    - ConcreteElement: implements accept() by calling the visitor's corresponding method

Double Dispatch:
    The visitor pattern uses double dispatch to invoke the correct operation.
    First dispatch: the element calls accept(visitor) -- polymorphic on the element type.
    Second dispatch: inside accept(), the element calls visitor.visit_X(self) -- polymorphic
    on the visitor type. This two-step indirection lets the runtime resolve both the element
    type AND the visitor type, achieving behavior that depends on both.
"""

from __future__ import annotations
from abc import ABC, abstractmethod


# ---------------------------------------------------------------------------
# Element hierarchy
# ---------------------------------------------------------------------------

class Element(ABC):
    """Base interface -- every document element must accept a visitor."""

    @abstractmethod
    def accept(self, visitor: ExportVisitor) -> str:
        ...


class Paragraph(Element):
    def __init__(self, text: str) -> None:
        self.text = text

    def accept(self, visitor: ExportVisitor) -> str:
        # Second dispatch: visitor now knows the concrete element type
        return visitor.visit_paragraph(self)


class Image(Element):
    def __init__(self, url: str, alt_text: str) -> None:
        self.url = url
        self.alt_text = alt_text

    def accept(self, visitor: ExportVisitor) -> str:
        return visitor.visit_image(self)


class Table(Element):
    def __init__(self, headers: list[str], rows: list[list[str]]) -> None:
        self.headers = headers
        self.rows = rows

    def accept(self, visitor: ExportVisitor) -> str:
        return visitor.visit_table(self)


# ---------------------------------------------------------------------------
# Visitor hierarchy
# ---------------------------------------------------------------------------

class ExportVisitor(ABC):
    """Each concrete visitor implements a different export format."""

    @abstractmethod
    def visit_paragraph(self, p: Paragraph) -> str:
        ...

    @abstractmethod
    def visit_image(self, img: Image) -> str:
        ...

    @abstractmethod
    def visit_table(self, tbl: Table) -> str:
        ...


class HtmlExporter(ExportVisitor):
    def visit_paragraph(self, p: Paragraph) -> str:
        return f"<p>{p.text}</p>"

    def visit_image(self, img: Image) -> str:
        return f'<img src="{img.url}" alt="{img.alt_text}" />'

    def visit_table(self, tbl: Table) -> str:
        header_cells = "".join(f"<th>{h}</th>" for h in tbl.headers)
        rows_html = ""
        for row in tbl.rows:
            cells = "".join(f"<td>{c}</td>" for c in row)
            rows_html += f"<tr>{cells}</tr>"
        return f"<table><tr>{header_cells}</tr>{rows_html}</table>"


class MarkdownExporter(ExportVisitor):
    def visit_paragraph(self, p: Paragraph) -> str:
        return p.text

    def visit_image(self, img: Image) -> str:
        return f"![{img.alt_text}]({img.url})"

    def visit_table(self, tbl: Table) -> str:
        header_line = "| " + " | ".join(tbl.headers) + " |"
        sep_line = "| " + " | ".join("---" for _ in tbl.headers) + " |"
        row_lines = "\n".join(
            "| " + " | ".join(row) + " |" for row in tbl.rows
        )
        return f"{header_line}\n{sep_line}\n{row_lines}"


class PlainTextExporter(ExportVisitor):
    def visit_paragraph(self, p: Paragraph) -> str:
        return p.text

    def visit_image(self, img: Image) -> str:
        return f"[Image: {img.alt_text} ({img.url})]"

    def visit_table(self, tbl: Table) -> str:
        col_widths = [len(h) for h in tbl.headers]
        for row in tbl.rows:
            for i, cell in enumerate(row):
                col_widths[i] = max(col_widths[i], len(cell))

        def fmt_row(cells: list[str]) -> str:
            return "  ".join(c.ljust(w) for c, w in zip(cells, col_widths))

        lines = [fmt_row(tbl.headers)]
        lines.append("  ".join("-" * w for w in col_widths))
        for row in tbl.rows:
            lines.append(fmt_row(row))
        return "\n".join(lines)


# ---------------------------------------------------------------------------
# Helper: export entire document
# ---------------------------------------------------------------------------

def export_document(elements: list[Element], visitor: ExportVisitor) -> str:
    """Apply the visitor to every element and join the results."""
    return "\n".join(el.accept(visitor) for el in elements)


# ---------------------------------------------------------------------------
# Main -- demonstrate exporting a document in three formats
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    # Build a small document
    document: list[Element] = [
        Paragraph("Welcome to the visitor pattern demo."),
        Image("https://example.com/photo.png", "A sample photo"),
        Table(
            headers=["Name", "Role"],
            rows=[
                ["Alice", "Engineer"],
                ["Bob", "Designer"],
            ],
        ),
    ]

    visitors = {
        "HTML": HtmlExporter(),
        "Markdown": MarkdownExporter(),
        "Plain Text": PlainTextExporter(),
    }

    for label, visitor in visitors.items():
        print(f"=== {label} Export ===")
        print(export_document(document, visitor))
        print()
